reports/reports_route.go

```diff
@@ -0,0 +1,141 @@
+			},
+			routes.QueryArgs{routes.AwsAccountQueryArg},
+			routes.QueryArgs{routes.ReportTypeQueryArg},
+			routes.QueryArgs{routes.FileNameQueryArg},
+		),
+	}.H().Register("/report")
+}
+
+func isUserAccount(tx *sql.Tx, user users.User, aa int) (bool, error) {
+	aaDB, err := models.AwsAccountByID(tx, aa)
+	return aaDB != nil && aaDB.UserID == user.Id, err
+}
+
+// getAwsReports returns the list of reports based on the query params, in JSON format.
+// The endpoint returns a list of strings following this format: report-type/file-name
+func getAwsReports(request *http.Request, a routes.Arguments) (int, interface{}) {
+	if config.ReportsBucket == "" {
+		return http.StatusInternalServerError, fmt.Errorf("Reports bucket not configured")
+	}
+	user := a[users.AuthenticatedUser].(users.User)
+	aa := a[routes.AwsAccountQueryArg].(int)
+	tx := a[db.Transaction].(*sql.Tx)
+	if aaOk, aaErr := isUserAccount(tx, user, aa); !aaOk {
+		return http.StatusUnauthorized, aaErr
+	}
+	svc := s3.New(awsSession.Session)
+	prefix := fmt.Sprintf("%d/", aa)
+	objects := []string{}
+	err := svc.ListObjectsPagesWithContext(request.Context(), &s3.ListObjectsInput{
+		Bucket: aws.String(config.ReportsBucket),
+		Prefix: aws.String(prefix),
+	}, func(p *s3.ListObjectsOutput, lastPage bool) bool {
+		for _, o := range p.Contents {
+			// Do not keep folders
+			if !strings.HasSuffix(*o.Key, "/") {
+				objects = append(objects, strings.TrimPrefix(aws.StringValue(o.Key), prefix))
+			}
+		}
+		return true // continue paging
+	})
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	return http.StatusOK, objects
+}
+
+type Report struct {
+	content []byte
+	name    string
+}
+
+func (r Report) GetFileContent() []byte {
+	return r.content
+}
+
+func (r Report) GetFileName() string {
+	return r.name
+}
+
+// getAwsReportsDownload returns the report based on the query params, in excel format.
+func getAwsReportsDownload(request *http.Request, a routes.Arguments) (int, interface{}) {
+	if config.ReportsBucket == "" {
+		return http.StatusInternalServerError, fmt.Errorf("Reports bucket not configured")
+	}
+	user := a[users.AuthenticatedUser].(users.User)
+	aa := a[routes.AwsAccountQueryArg].(int)
+	reportType := a[routes.ReportTypeQueryArg].(string)
+	reportName := a[routes.FileNameQueryArg].(string)
+	tx := a[db.Transaction].(*sql.Tx)
+	if aaOk, aaErr := isUserAccount(tx, user, aa); !aaOk {
+		return http.StatusUnauthorized, aaErr
+	}
+	reportPath := path.Join(strconv.Itoa(aa), reportType, reportName)
+	buff := &aws.WriteAtBuffer{}
+	downloader := s3manager.NewDownloader(awsSession.Session)
+	_, err := downloader.Download(buff,
+		&s3.GetObjectInput{
+			Bucket: aws.String(config.ReportsBucket),
+			Key:    aws.String(reportPath),
+		})
+	if err != nil {
+		return http.StatusNotFound, fmt.Errorf("The specified key does not exist")
+	}
+	return http.StatusOK, Report{buff.Bytes(), reportName}
```

Comment:
Does that buffer the file? The report could be really huge.

Reply:
@thibautcornolti Yes it does but I'm not sure there's a better way to do it. For now the biggest report I generated was 83kb so not really an issue

Reply:
If it does, it's good to me.